pydoctor/astbuilder.py

```diff
@@ -512,36 +485,81 @@ def _annotation_from_attrib(self, expr, ctx):
                 return _infer_type(default)
         return None
 
-    def _unstring_annotation(self, node):
+    def _annotations_from_function(self, func: ast.FunctionDef) -> Mapping[str, ast.expr]:
+        """Get annotations from a function definition.
+        @param func: The function definition's AST.
+        @return: Mapping from argument name to annotation.
+            The name C{return} is used for the return type.
+            Unannotated arguments are omitted.
+        """
+        def _get_all_args() -> Iterator[ast.arg]:
+            base_args = func.args
+            # New on Python 3.8 -- handle absence gracefully
+            try:
+                yield from base_args.posonlyargs
+            except AttributeError:
+                pass
+            yield from base_args.args
+            varargs = base_args.vararg
+            if varargs:
+                yield varargs
+            yield from base_args.kwonlyargs
+            kwargs = base_args.kwarg
+            if kwargs:
+                yield kwargs
+        def _get_all_ast_annotations() -> Iterator[Tuple[str, ast.expr]]:
+            for arg in _get_all_args():
+                if arg.annotation:
+                    yield arg.arg, arg.annotation
+            returns = func.returns
+            if returns:
+                yield 'return', returns
+        return {name: self._unstring_annotation(value)
+                for name, value in _get_all_ast_annotations()}
+
+    def _unstring_annotation(self, node: ast.expr) -> ast.expr:
         """Replace all strings in the given expression by parsed versions.
-        Returns the resulting node, or None if parsing failed.
+        @return: The unstringed node. If parsing fails, an error is logged
+            and the original node is returned.
         """
         try:
-            return _AnnotationStringParser().visit(node)
+            expr = _AnnotationStringParser().visit(node)
+            assert isinstance(expr, ast.expr), expr
+            return expr
         except SyntaxError as ex:
             self.builder.currentMod.report(
                     f'syntax error in annotation: {ex}',
                     lineno_offset=node.lineno)
-            return None
```

Comment:
This won't trigger for the `AssertionError`, so is this supposed to be triggered from the `expr = _Annotation...` line? In this case, let's have the `except` only around the line and use `else`.

Reply:
Yes, it will only catch `SyntaxError` raised by `_AnnotationStringParser`. We can indeed use an `else:` clause; if that makes the code easier to read, it's worth one extra line.